feature.ltm-ifiles

jasonrahm opened the pull request:
Problem: Adds ifiles support

f5-rahm committed:
Fixes #536

Problem: Adds ifiles support

pjbreaux commented:
@jasonrahm: Could you look to add some functional tests for this feature. Let me know if you need some answers.

zancas commented:
@jasonrahm To create a test object that acts like a file:

``` Python
import StringIO

fakefileobject = StringIO.StringIO('MYFAKEFILEDATA')

```

f5-rahm committed:
Fixes #536

Problem:functional test for /test/functional/tm/sys/test_file.py

Isn't working, getting 'file' object has no attribute 'rfind'

zancas commented on test/functional/tm/sys/test_file.py:
```diff
@@ -0,0 +1,55 @@
+TESTDESCRIPTION = "TESTDESCRIPTION"
+
+
+def setup_ifile_test(request, bigip, name, sourcepath):
```
replace `bigip` with `mgmt_root`

f5-rahm committed:
Fixes #536

Problem:functional test for /test/functional/tm/sys/test_file.py

Haven't t

zancas commented on test/functional/tm/sys/test_file.py:
```diff
@@ -0,0 +1,55 @@
+                raise
+    request.addfinalizer(teardown)
+
+    if1 = bigip.sys.file.ifiles.ifile.create(name=name, sourcePath=sourcepath)
```
You'll need to change this line:

`bigip` --> `mgmt_root.tm`

zancas commented on test/functional/tm/sys/test_file.py:
```diff
@@ -0,0 +1,55 @@
+        ntf_basename = os.path.basename(ntf.name)
+        ntf.write('this is a test file')
+        #Upload the file
+        bigip.shared.file_transfer.uploads.upload_file(ntf)
```
Change this:

`bigip` --> `mgmt_root`

zancas commented on test/functional/tm/sys/test_file.py:
```diff
@@ -0,0 +1,55 @@
+        assert if1.name == ntf_basename
+
+        # Load - Test with the various partition/name combinations
+        if2 = bigip.sys.file.ifiles.ifile.load(name='testifileobject')
```
`bigip` --> `mgmt_root.tm`

zancas commented on test/functional/tm/sys/test_file.py:
```diff
@@ -0,0 +1,55 @@
+    return if1
+
+
+class Test_iFile(object):
```
Delete the class, just use `def test_CURDL(request, mgmt_root)` at the top level of the module.

f5-rahm committed:
Fixes #536

tests added for:
   /f5/bigip/tm/sys/test
   /f5/bigip/tm/ltm/test
   /test/functional/tm/sys

   Tests passing at this time

f5-rahm committed:
Fixes #536

tests added for:
   /f5/bigip/tm/sys/test
   /f5/bigip/tm/ltm/test
   /test/functional/tm/sys

   Tests passing at this time; updated test to remove class

f5-rahm committed:
Fixes #536

tests added for:
   /f5/bigip/tm/sys/test
   /f5/bigip/tm/ltm/test
   /test/functional/tm/sys

   Tests passing at this time; updated test to restore class

zancas commented:
@jasonrahm I've pulled and rebased your branch on top of latest development.

The good news is that (in my local environment) all the _other_ tests (that were passing before) still pass.

zancas commented:
File creation appears to succeed which is good news for our upload logic.

However there's an unmet expectation on the part of the test code, that a particular directory will exist in `/var/tmp/tmsh`   that random-name-looking directory is expected to have a `data` file.  Other examples of that file that I've seen appear to be logs of tmsh commands?

Here's the error output I see:

```
clear && py.test -s -v --bigip REDACTED --username admin --password admin --ignore=test/functional/tm/gtm/ test/functional/tm/sys/test_file.py::test_CURDL
================================================================ test session starts ================================================================
platform linux2 -- Python 2.7.12, pytest-2.8.7, py-1.4.31, pluggy-0.3.1 -- /usr/bin/python
cachedir: .cache
rootdir: /root/devenv/f5-common-python, inifile:
plugins: cov-2.2.1, symbols-0.1.0a3
collected 2 items

test/functional/tm/sys/test_file.py::test_CURDL DEBUG:root:get WITH uri: https://10.190.7.130:443/mgmt/tm/sys/ AND suffix:  AND kwargs: {}
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): 10.190.7.130
DEBUG:requests.packages.urllib3.connectionpool:"GET /mgmt/tm/sys/ HTTP/1.1" 200 2952
DEBUG:root:RESPONSE::STATUS: 200 Content-Type: application/json Content-Encoding: None
Text: u'{"kind":"tm:sys:syscollectionstate","selfLink":"https://localhost/mgmt/tm/sys?ver=11.6.0","items":[{"reference":{"link":"https://localhost/mgmt/tm/sys/application?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/crypto?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/daemon-log-settings?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/disk?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/file?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/fpga?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/icall?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/ipfix?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/log-config?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/sflow?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/software?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/url-db?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/cluster?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/config?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/daemon-ha?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/datastor?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/db?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/dns?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/feature-module?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/folder?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/global-settings?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/ha-group?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/httpd?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/log-rotate?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/management-dhcp?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/management-ip?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/management-route?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/ntp?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/outbound-smtp?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/provision?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/scriptd?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/service?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/smtp-server?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/snmp?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/sshd?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/state-mirroring?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/syslog?ver=11.6.0"}},{"reference":{"link":"https://localhost/mgmt/tm/sys/ucs?ver=11.6.0"}}]}'
DEBUG:root:{'headers': {'Content-Range': '0-18/19', 'Content-Type': 'application/octet-stream'}, 'data': 'this is a test file', 'verify': False}
DEBUG:root:post WITH uri: https://REDACTED:443/mgmt/shared/file-transfer/uploads/tmpeJSXhx AND suffix:  AND kwargs: {'headers': {'Content-Range': '0-18/19', 'Content-Type': 'application/octet-stream'}, 'data': 'this is a test file', 'verify': False}
DEBUG:requests.packages.urllib3.connectionpool:"POST /mgmt/shared/file-transfer/uploads/tmpeJSXhx HTTP/1.1" 200 235
DEBUG:root:RESPONSE::STATUS: 200 Content-Type: application/json Content-Encoding: None
Text: u'{"remainingByteCount":0,"usedChunks":{"0":19},"totalByteCount":19,"localFilePath":"/var/config/rest/downloads/tmpeJSXhx","temporaryFilePath":"/var/config/rest/downloads/tmp/tmpeJSXhx","generation":0,"lastUpdateMicros":1470244580305047}'
DEBUG:root:post WITH uri: https://REDACTED:443/mgmt/tm/sys/file/ifile/ AND suffix:  AND kwargs: {'json': {'sourcePath': 'file:/var/config/rest/downloads/tmpeJSXhx', 'name': 'tmpeJSXhx'}}
DEBUG:requests.packages.urllib3.connectionpool:"POST /mgmt/tm/sys/file/ifile/ HTTP/1.1" 400 104
DEBUG:root:RESPONSE::STATUS: 400 Content-Type: application/json Content-Encoding: None
Text: u'{"code":400,"message":"01070712:3: file (/var/tmp/tmsh/macaDr/data) expected to exist.","errorStack":[]}'
FAILED

===================================================================== FAILURES ======================================================================
____________________________________________________________________ test_CURDL _____________________________________________________________________

request = <FixtureRequest for <Function 'test_CURDL'>>, mgmt_root = <f5.bigip.ManagementRoot object at 0x7f26e4d0b250>

    def test_CURDL(request, mgmt_root):
        # Create
        ntf = NamedTemporaryFile()
        ntf_basename = os.path.basename(ntf.name)
        ntf.write('this is a test file')
        ntf.seek(0)
        #Upload the file
        mgmt_root.shared.file_transfer.uploads.upload_file(ntf.name)

        tpath_name = 'file:/var/config/rest/downloads/{0}'.format(ntf_basename)
>       if1 = setup_ifile_test(request, mgmt_root, ntf_basename, tpath_name)

test/functional/tm/sys/test_file.py:51:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test/functional/tm/sys/test_file.py:26: in setup_ifile_test
    sourcePath=sourcepath)
f5/bigip/resource.py:842: in create
    return self._create(**kwargs)
f5/bigip/resource.py:809: in _create
    response = session.post(_create_uri, json=kwargs, **requests_params)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <icontrol.session.iControlRESTSession object at 0x7f26e4d0b450>, RIC_base_uri = 'https://REDACTED:443/mgmt/tm/sys/file/ifile/'
kwargs = {'json': {'name': 'tmpeJSXhx', 'sourcePath': 'file:/var/config/rest/downloads/tmpeJSXhx'}}, partition = '', name = '', suffix = ''
uri_as_parts = False, REST_uri = 'https://10.190.7.130:443/mgmt/tm/sys/file/ifile/'
pre_message = "post WITH uri: https://10.190.7.130:443/mgmt/tm/sys/file/ifile/ AND suffix:  AND kwargs: {'json': {'sourcePath': 'file:/var/config/rest/downloads/tmpeJSXhx', 'name': 'tmpeJSXhx'}}"
response = <Response [400]>

    @functools.wraps(method)
    def wrapper(self, RIC_base_uri, **kwargs):
        partition = kwargs.pop('partition', '')
        name = kwargs.pop('name', '')
        suffix = kwargs.pop('suffix', '')
        uri_as_parts = kwargs.pop('uri_as_parts', False)
        if uri_as_parts:
            REST_uri = generate_bigip_uri(RIC_base_uri, partition, name,
                                          suffix, **kwargs)
        else:
            REST_uri = RIC_base_uri
        pre_message = "%s WITH uri: %s AND suffix: %s AND kwargs: %s" %\
            (method.__name__, REST_uri, suffix, kwargs)
        logging.debug(pre_message)
        response = method(self, REST_uri, **kwargs)
        post_message =\
            "RESPONSE::STATUS: %s Content-Type: %s Content-Encoding:"\
            " %s\nText: %r" % (response.status_code,
                               response.headers.get('Content-Type', None),
                               response.headers.get('Content-Encoding', None),
                               response.text)
        logging.debug(post_message)
        if response.status_code not in range(200, 207):
            error_message = '%s Unexpected Error: %s for uri: %s\nText: %r' %\
                            (response.status_code,
                             response.reason,
                             response.url,
                             response.text)
>           raise iControlUnexpectedHTTPError(error_message, response=response)
E           iControlUnexpectedHTTPError: 400 Unexpected Error: Bad Request for uri: https://REDACTED:443/mgmt/tm/sys/file/ifile/
E           Text: u'{"code":400,"message":"01070712:3: file (/var/tmp/tmsh/macaDr/data) expected to exist.","errorStack":[]}'

/usr/local/lib/python2.7/dist-packages/icontrol/session.py:230: iControlUnexpectedHTTPError
============================================================= 1 failed in 0.37 seconds ==============================================================
```

zancas commented:
Forehead smack...

zancas commented:
OK, according to the tmsh help docs the "source-path" (which should be the same as the REST key "sourcePath") has to be a URI:

```
       source-path [URL]
            This attribute takes a URL, for example:

                 source-path http://file-server/ifiles/AUL_1.cls

                 source-path https://file-server/ifiles/CNN.x

                 source-path ftp://username:password@server/ifiles/latest.class
```

zancas commented:
Here's my version of the feature:

https://github.com/zancas/f5-common-python/tree/feature.ltm-ifiles_536

Feel free to use it, reset to it, whatever.

jasonrahm commented:
The tpath is inline in my version, maybe the line wrapping I was doing was messing up. I’ll update to yours and commit.

Jason

From: Za Wilgustus notifications@github.com
Reply-To: F5Networks/f5-common-python reply@reply.github.com
Date: Wednesday, August 3, 2016 at 1:11 PM
To: F5Networks/f5-common-python f5-common-python@noreply.github.com
Cc: Jason Rahm J.Rahm@F5.com, Mention mention@noreply.github.com
Subject: Re: [F5Networks/f5-common-python] Fixes #536 (#537)

Here's my version of the feature:

https://github.com/zancas/f5-common-python/tree/feature.ltm-ifiles_536

Feel free to use it, reset to it, whatever.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/F5Networks/f5-common-python/pull/537#issuecomment-237319056, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AA0SSx0w1koW_MjzaJiEdyZuNDv7BXe3ks5qcNnZgaJpZM4JGgmF.

f5-rahm committed:
Fixes #536

tests added for:
   /f5/bigip/tm/sys/test
   /f5/bigip/tm/ltm/test
   /test/functional/tm/sys

   Tests passing at this time; updated test to restore class

f5-rahm committed:
Fixes #536

tests added for:
   /f5/bigip/tm/sys/test
   /f5/bigip/tm/ltm/test
   /test/functional/tm/sys

   Tests passing at this time; updated test to restore class

f5-rahm committed:
Fixes #536

tests added for:
   /f5/bigip/tm/sys/test
   /f5/bigip/tm/ltm/test
   /test/functional/tm/sys

   Tests passing at this time; updated test to restore class

zancas commented:
Hmm...   some checks still fail.

jasonrahm commented:
locally everything is passing, but not sure what the problem is.

src: mobile

On Aug 3, 2016, at 3:50 PM, Za Wilgustus <notifications@github.com<mailto:notifications@github.com>> wrote:

Hmm... some checks still fail.

## 

You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/F5Networks/f5-common-python/pull/537#issuecomment-237368729, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AA0SS_YUfI-je_-3TA-aba_J1lF4QdD6ks5qcP7cgaJpZM4JGgmF.

caphrim007 commented:
@jasonrahm it's flake8 errors here

```
./test/functional/tm/sys/test_file.py:18:1: H306  imports not in alphabetical order (tempfile.namedtemporaryfile, os)
./test/functional/tm/sys/test_file.py:32:1: H405  multi line docstring summary not separated with an empty line
```

jasonrahm commented:
Gah! Will fix tonight

src: mobile

On Aug 3, 2016, at 5:02 PM, Tim Rupp <notifications@github.com<mailto:notifications@github.com>> wrote:

@jasonrahmhttps://github.com/jasonrahm it's flake8 errors here

./test/functional/tm/sys/test_file.py:18:1: H306  imports not in alphabetical order (tempfile.namedtemporaryfile, os)
./test/functional/tm/sys/test_file.py:32:1: H405  multi line docstring summary not separated with an empty line

## 

You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/F5Networks/f5-common-python/pull/537#issuecomment-237388946, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AA0SSxHMXOgRZ71URPBFPB-4IhSt1ypOks5qcQ_6gaJpZM4JGgmF.

f5-rahm committed:
Fixes #536

tests added for:
   /f5/bigip/tm/sys/test
   /f5/bigip/tm/ltm/test
   /test/functional/tm/sys

   Tests passing at this time; updated test to restore class

zancas commented:
My local tests fail with the same error I had before.
Also we need to purge print statements before we can merge.

jasonrahm commented:
What version tmos? I'm on 12.1 maybe a version issue? I'll pull print statements

src: mobile

On Aug 3, 2016, at 7:29 PM, Za Wilgustus <notifications@github.com<mailto:notifications@github.com>> wrote:

My local tests fail with the same error I had before.
Also we need to purge print statements before we can merge.

## 

You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/F5Networks/f5-common-python/pull/537#issuecomment-237416246, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AA0SS08NAhx7XAAcRnmp_MkP64KifBEPks5qcTJRgaJpZM4JGgmF.

zancas commented:
OK, we need to remove that print statement I added.
Then let's think about testing the other operations the resource should support.

zancas commented:
Refresh, Update, Delete, Exists?

pjbreaux commented:
Oh yeah @jasonrahm, one thing to think about is squashing some of your commits into meaningful chunks (as meaningful as they can be). We have no hard rule about the number of commits to merge, but we often find that the commits can be bunched together to form a few coherent blocks of changes. We can certainly merge this one without bunching those together, but just keep it in mind for future features. Also, if you need some help squashing or rebasing interactively, just shout out here.

jasonrahm commented:
Squashing commits would be good, any advice appreciated

src: mobile

On Aug 3, 2016, at 9:19 PM, pjbreaux <notifications@github.com<mailto:notifications@github.com>> wrote:

Oh yeah @jasonrahmhttps://github.com/jasonrahm, one thing to think about is squashing some of your commits into meaningful chunks (as meaningful as they can be). We have no hard rule about the number of commits to merge, but we often find that the commits can be bunched together to form a few coherent blocks of changes. We can certainly merge this one without bunching those together, but just keep it in mind for future features. Also, if you need some help squashing or rebasing interactively, just shout out here.

## 

You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/F5Networks/f5-common-python/pull/537#issuecomment-237433938, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AA0SS5cf6At7PI5LIwXa96aXpLxTT_Vjks5qcUxDgaJpZM4JGgmF.

caphrim007 commented:
@pjbreaux we do have the ability to squash commits at merge time

pjbreaux commented:
This gives you a pretty quick run-through: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#Changing-Multiple-Commit-Messages

Generally, what we do is do a `git log --decorate --oneline` and find the commit farthest back that we want to squash (this means only the changes we've made locally in our fork and haven't been merged yet). We copy the hash of that first commit when we started working on the feature and do `git rebase -i <copied_hash>`. This brings us into interactive mode where we can pick, edit, fixup etc.... And remember, if you remove a line during interactive rebasing, it will pull that commit completely out of the rebase (generally no good, unless you meant to do it).

And I haven't tried what @tar suggested above. That might be a good solution today.

caphrim007 commented:
@pjbreaux if it's ready to merge i can try the squash button in the UI. I've never actually used it. this could be a good chance to try it

pjbreaux commented:
Not sure if it's ready. We'll get the go-ahead from @jasonrahm and @zancas

jasonrahm commented:
I think Za wanted me to work on more test cases, so I'll take a look at that and the resources below. Thanks all. Heading back from dinner now and not sure how late I'll last tonight.

src: mobile

On Aug 3, 2016, at 10:24 PM, pjbreaux <notifications@github.com<mailto:notifications@github.com>> wrote:

Not sure if it's ready. We'll get the go-ahead from @jasonrahmhttps://github.com/jasonrahm and @zancashttps://github.com/zancas

## 

You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/F5Networks/f5-common-python/pull/537#issuecomment-237442372, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AA0SS6FD2n82S3kcW39OoSglKKhqPFtMks5qcVtYgaJpZM4JGgmF.

f5-rahm committed:
Fixes #536 ; added refresh and update tests

jasonrahm commented:
Might be good in this case to hit the squash button, I tried to rebase locally and got an error. I added some extra assertions for testing updates and refreshes. Hopefully that’ll get this over the hump.

Jason

From: Tim Rupp notifications@github.com
Reply-To: F5Networks/f5-common-python reply@reply.github.com
Date: Wednesday, August 3, 2016 at 10:21 PM
To: F5Networks/f5-common-python f5-common-python@noreply.github.com
Cc: Jason Rahm J.Rahm@F5.com, Mention mention@noreply.github.com
Subject: Re: [F5Networks/f5-common-python] Fixes #536 (#537)

@pjbreauxhttps://github.com/pjbreaux if it's ready to merge i can try the squash button in the UI. I've never actually used it. this could be a good chance to try it

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/F5Networks/f5-common-python/pull/537#issuecomment-237442066, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AA0SS98eoK9u57IRTXyzen6xQli80Xevks5qcVq6gaJpZM4JGgmF.

f5-rahm committed:
Fixes #536 ; added IFILE fixture and ltm/test_ifile tests.

f5-rahm committed:
Fixes #536 ; fixed import ordering
